File: transcribe.py
import hashlib
import hmac
import base64
import time
import random
import string
import requests
import json
import os
import subprocess
from datetime import datetime, timezone, timedelta
from urllib.parse import quote

from config import XFYUN_APP_ID, XFYUN_API_KEY, XFYUN_API_SECRET

APP_ID            = XFYUN_APP_ID
ACCESS_KEY_ID     = XFYUN_API_KEY
ACCESS_KEY_SECRET = XFYUN_API_SECRET

UPLOAD_URL = "https://office-api-ist-dx.iflyaisol.com/v2/upload"
RESULT_URL = "https://office-api-ist-dx.iflyaisol.com/v2/getResult"
AUDIO_FILE = "/home/user/Interview_notes/小米创新AI产品经理.m4a"


def get_datetime():
    """返回 yyyy-MM-ddTHH:mm:ss+0800 格式的时间"""
    tz_cst = timezone(timedelta(hours=8))
    return datetime.now(tz_cst).strftime("%Y-%m-%dT%H:%M:%S+0800")


def rand_str(n=16):
    return ''.join(random.choices(string.ascii_letters + string.digits, k=n))


def calc_signature(params: dict) -> str:
    """
    签名算法：
    1. 排除 signature 字段，其余参数按 key 自然排序
    2. 对每个 k=v URL编码后用 & 拼接
    3. HMAC-SHA1(baseString, ACCESS_KEY_SECRET) → Base64
    """
    sorted_pairs = sorted(
        [(k, v) for k, v in params.items() if k != "signature" and v is not None and v != ""],
        key=lambda x: x[0]
    )
    base_string = "&".join(
        f"{quote(str(k), safe='')}={quote(str(v), safe='')}"
        for k, v in sorted_pairs
    )
    sig = base64.b64encode(
        hmac.new(
            ACCESS_KEY_SECRET.encode("utf-8"),
            base_string.encode("utf-8"),
            hashlib.sha1
        ).digest()
    ).decode()
    return sig


def build_params(extra: dict) -> tuple[dict, str]:
    """返回 (query_params, signature)，signature 放请求头"""
    params = {
        "appId": APP_ID,
        "accessKeyId": ACCESS_KEY_ID,
        "dateTime": get_datetime(),
        "signatureRandom": rand_str(16),
    }
    params.update(extra)
    signature = calc_signature(params)
    return params, signature


def get_duration_ms(path):
    result = subprocess.run(
        ["ffprobe", "-v", "quiet", "-print_format", "json", "-show_format", path],
        capture_output=True, text=True
    )
    data = json.loads(result.stdout)
    return int(float(data["format"]["duration"]) * 1000)


def main():
    file_name    = os.path.basename(AUDIO_FILE)
    file_size    = os.path.getsize(AUDIO_FILE)
    duration_ms  = get_duration_ms(AUDIO_FILE)
    print(f"文件: {file_name}  大小: {file_size/1024/1024:.1f}MB  时长: {duration_ms/1000:.1f}s")

    # ── Step 1: 上传 ─────────────────────────────────────────────
    params, sig = build_params({
        "fileName": file_name,
        "fileSize": str(file_size),
        "duration": str(duration_ms),
        "language": "autodialect",
        "roleType": "1",          # 开启说话人分离（2人）
    })

    print("上传中...")
    with open(AUDIO_FILE, "rb") as f:
        resp = requests.post(
            UPLOAD_URL,
            params=params,
            data=f,
            headers={"Content-Type": "application/octet-stream", "signature": sig},
        )

    print(f"HTTP {resp.status_code}")
    upload_result = resp.json()
    print(f"[upload] {json.dumps(upload_result, ensure_ascii=False)}")

    if upload_result.get("code") != "000000":
        print("上传失败，退出")
        return

    order_id     = upload_result["content"]["orderId"]
    estimate_ms  = upload_result["content"].get("taskEstimateTime", 0)
    print(f"orderId: {order_id}  预计: {estimate_ms/1000:.0f}s")

    # ── Step 2: 轮询结果 ─────────────────────────────────────────
    print("\n等待转写（每15秒查询一次）...")
    for attempt in range(100):
        time.sleep(15)
        params, sig = build_params({"orderId": order_id, "resultType": "transfer"})
        resp = requests.post(RESULT_URL, params=params, headers={"signature": sig})
        result = resp.json()
        code = result.get("code")
        order_info = result.get("content", {}).get("orderInfo", {})
        status = order_info.get("status")
        print(f"  [{attempt+1}] code={code} status={status}")

        if code == "000000" and status == 4:
            print("转写完成！")
            break
        if code not in ("000000", "26682"):
            print(f"异常: {json.dumps(result, ensure_ascii=False)}")
            return
    else:
        print("超时退出")
        return

    # ── Step 3: 解析结果 ─────────────────────────────────────────
    order_result_raw = result.get("content", {}).get("orderResult", "{}")
    if isinstance(order_result_raw, str):
        order_result = json.loads(order_result_raw)
    else:
        order_result = order_result_raw

    lattice = order_result.get("lattice", [])

    structured = []
    for item in lattice:
        json_1best = item.get("json_1best", "{}")
        if isinstance(json_1best, str):
            seg = json.loads(json_1best)
        else:
            seg = json_1best

        st = seg.get("st", {})
        bg = int(st.get("bg", 0))
        ed = int(st.get("ed", 0))
        speaker = st.get("rl", 0)

        text = ""
        for rt in st.get("rt", []):
            for ws in rt.get("ws", []):
                for cw in ws.get("cw", []):
                    text += cw.get("w", "")

        if text.strip():
            structured.append({
                "speaker": speaker,
                "start_ms": bg,
                "end_ms": ed,
                "text": text,
            })

    output_path = AUDIO_FILE.replace(".m4a", "_transcript.json")
    with open(output_path, "w", encoding="utf-8") as f:
        json.dump(structured, f, ensure_ascii=False, indent=2)
    print(f"\n结果已保存: {output_path}")

    print("\n===== 转写结果 =====")
    for item in structured:
        t_s = item["start_ms"] / 1000
        t_e = item["end_ms"] / 1000
        print(f"[Speaker {item['speaker']}] {t_s:.1f}s~{t_e:.1f}s: {item['text']}")


if __name__ == "__main__":
    main()

File: interview-analysis-report/server/transcribe.ts
import crypto from "crypto";
import fs from "fs";
import path from "path";
import { execSync } from "child_process";
import dotenv from "dotenv";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

dotenv.config({ path: path.resolve(__dirname, "../.env") });

const XFYUN_APP_ID = process.env.XFYUN_APP_ID!;
const XFYUN_API_KEY = process.env.XFYUN_API_KEY!;
const XFYUN_API_SECRET = process.env.XFYUN_API_SECRET!;

const UPLOAD_URL = "https://office-api-ist-dx.iflyaisol.com/v2/upload";
const RESULT_URL = "https://office-api-ist-dx.iflyaisol.com/v2/getResult";

export interface TranscriptSegment {
  speaker: string;
  start_ms: number;
  end_ms: number;
  text: string;
}

/** Strict URL-encode: also encodes !'()* to match Python's quote(safe='') */
function strictEncode(str: string): string {
  return encodeURIComponent(String(str)).replace(
    /[!'()*]/g,
    (c) => "%" + c.charCodeAt(0).toString(16).toUpperCase()
  );
}

function xfyunDateTime(): string {
  const now = new Date();
  const offset = 8 * 60; // UTC+8
  const local = new Date(now.getTime() + offset * 60 * 1000);
  const y = local.getUTCFullYear();
  const M = String(local.getUTCMonth() + 1).padStart(2, "0");
  const d = String(local.getUTCDate()).padStart(2, "0");
  const h = String(local.getUTCHours()).padStart(2, "0");
  const m = String(local.getUTCMinutes()).padStart(2, "0");
  const s = String(local.getUTCSeconds()).padStart(2, "0");
  return `${y}-${M}-${d}T${h}:${m}:${s}+0800`;
}

function xfyunRand(n = 16): string {
  const chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789";
  let result = "";
  for (let i = 0; i < n; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

function xfyunSignature(params: Record<string, string>): string {
  const pairs = Object.entries(params)
    .filter(([k, v]) => k !== "signature" && v)
    .sort(([a], [b]) => a.localeCompare(b));

  const baseString = pairs
    .map(([k, v]) => `${strictEncode(k)}=${strictEncode(v)}`)
    .join("&");

  return crypto
    .createHmac("sha1", XFYUN_API_SECRET)
    .update(baseString)
    .digest("base64");
}

function xfyunParams(extra: Record<string, string>): { params: Record<string, string>; signature: string } {
  const params: Record<string, string> = {
    appId: XFYUN_APP_ID,
    accessKeyId: XFYUN_API_KEY,
    dateTime: xfyunDateTime(),
    signatureRandom: xfyunRand(),
    ...extra,
  };
  const signature = xfyunSignature(params);
  return { params, signature };
}

export function getDurationMs(filePath: string): number {
  const result = execSync(
    `ffprobe -v quiet -print_format json -show_format "${filePath}"`,
    { encoding: "utf-8" }
  );
  const data = JSON.parse(result);
  return Math.floor(parseFloat(data.format.duration) * 1000);
}

async function uploadAudio(audioPath: string): Promise<{ orderId: string; estimate: number }> {
  const fileName = path.basename(audioPath);
  const fileSize = fs.statSync(audioPath).size;
  const durationMs = getDurationMs(audioPath);

  console.log(`  文件: ${fileName}  大小: ${(fileSize / 1024 / 1024).toFixed(1)}MB  时长: ${Math.floor(durationMs / 1000)}s`);

  const { params, signature } = xfyunParams({
    fileName,
    fileSize: String(fileSize),
    duration: String(durationMs),
    language: "autodialect",
    roleType: "1",
  });

  const queryString = Object.entries(params)
    .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
    .join("&");

  const body = fs.readFileSync(audioPath);

  const resp = await fetch(`${UPLOAD_URL}?${queryString}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/octet-stream",
      signature,
    },
    body,
  });

  const uploadResult = await resp.json();
  if (uploadResult.code !== "000000") {
    throw new Error(`上传失败: ${JSON.stringify(uploadResult)}`);
  }

  const orderId = uploadResult.content.orderId;
  const estimate = (uploadResult.content.taskEstimateTime || 0) / 1000;
  console.log(`  orderId: ${orderId}  预计: ${Math.floor(estimate)}s`);

  return { orderId, estimate };
}

async function pollResult(
  orderId: string,
  onProgress?: (attempt: number, maxAttempts: number) => void
): Promise<any> {
  const maxAttempts = 100;

  for (let attempt = 0; attempt < maxAttempts; attempt++) {
    await new Promise((resolve) => setTimeout(resolve, 15000));

    const { params, signature } = xfyunParams({
      orderId,
      resultType: "transfer",
    });

    const queryString = Object.entries(params)
      .map(([k, v]) => `${encodeURIComponent(k)}=${encodeURIComponent(v)}`)
      .join("&");

    const resp = await fetch(`${RESULT_URL}?${queryString}`, {
      method: "POST",
      headers: { signature },
    });

    const result = await resp.json();
    const code = result.code;
    const status = result.content?.orderInfo?.status;

    onProgress?.(attempt + 1, maxAttempts);

    if (code === "000000" && status === 4) {
      console.log("  转写完成!");
      return result;
    }

    if (code !== "000000" && code !== "26682") {
      throw new Error(`转写异常: ${JSON.stringify(result)}`);
    }
  }

  throw new Error("转写超时（超过100次轮询）");
}

function parseResult(result: any): TranscriptSegment[] {
  let orderResult = result.content.orderResult;
  if (typeof orderResult === "string") {
    orderResult = JSON.parse(orderResult);
  }

  const structured: TranscriptSegment[] = [];

  for (const item of orderResult.lattice || []) {
    const seg = JSON.parse(item.json_1best || "{}");
    const st = seg.st || {};

    const text = (st.rt || [])
      .flatMap((rt: any) => rt.ws || [])
      .flatMap((ws: any) => ws.cw || [])
      .map((cw: any) => cw.w || "")
      .join("");

    if (text.trim()) {
      structured.push({
        speaker: st.rl || "0",
        start_ms: parseInt(st.bg || "0", 10),
        end_ms: parseInt(st.ed || "0", 10),
        text,
      });
    }
  }

  console.log(`  转写片段数: ${structured.length}`);
  return structured;
}

export async function transcribe(
  audioPath: string,
  onProgress?: (stage: string, detail?: string, percent?: number) => void
): Promise<TranscriptSegment[]> {
  onProgress?.("uploading", "正在上传音频文件...", 10);
  const { orderId } = await uploadAudio(audioPath);

  onProgress?.("transcribing", "正在转写，请等待...", 20);
  const result = await pollResult(orderId, (attempt, max) => {
    // 20% ~ 90% maps to polling progress
    const pct = Math.round(20 + (attempt / max) * 70);
    onProgress?.("transcribing", `轮询中 (${attempt}/${max})...`, pct);
  });

  onProgress?.("parsing", "正在解析结果...", 95);
  return parseResult(result);
}
